File: practiced/PracticeInterface.java
package com.practice.practiceinterface;

import static java.lang.System.out;

interface IntSequence {
    boolean hasNext();
    int next();
}

class PracticeInterface {

    public static void main(String[] args) {
        IntSequence intSeq = new SquareSequence();
        var avg = getAvearge(intSeq, 100);
        out.println(avg);
        //
        IntSequence digSeq = new DigitSequence(1978);
        var digAvg = getAvearge(digSeq, 100);
        out.println(digAvg);
    }

    static double getAvearge(IntSequence seq, int n) {
        int count = 0;
        double sum = 0;
        while (seq.hasNext() && count < n) {
            sum += seq.next();
            count++;
        }
        return sum/count;
    }
}

class SquareSequence implements IntSequence {
    private int number;

    @Override
    public boolean hasNext() {
        return true;
    }

    @Override
    public int next() {
        number++;
        return number * number;
    }
}
class DigitSequence implements IntSequence {
    private int digits;
    DigitSequence(int digits) {
        this.digits = digits;
    }

    public boolean hasNext() {
        return this.digits != 0;
    }

    public int next() {
        int result = this.digits % 10;
        this.digits /= 10;
        return result;
    }
}

File: practiced/PracticeInterface.scala
package com.progscala3.practiceinterface

trait IntSequence:
  def hasNext: Boolean = true
  def next: Int;

package calculate:
  def getAverage(seq: IntSequence, count: Int): Double =
    var counter = 0
    var sum: Double = 0
    while seq.hasNext && counter < count do
      sum += seq.next
      counter += 1
    sum / counter

package otherClasses:
  class SquareSequence(var number: Int = 0) extends IntSequence:
    override def next: Int =
      number += 1
      number * number

  class DigitSequence(var number: Int) extends IntSequence:
    override def hasNext: Boolean = number != 0
    override def next: Int =
      val result = number % 10
      number /= 10
      result

@main def main(args: String*): Unit =
  import otherClasses.*
  import calculate.*

  val intSeq = SquareSequence()
  println(getAverage(intSeq, 100))
  //
  val digSeq = new DigitSequence(1978);
  val digAvg = getAverage(digSeq, 100);
  println(digAvg);
